src/Features/Core/Portable/CodeCleanup/AbstractCodeCleanupService.cs

```diff
@@ -136,7 +156,49 @@ protected AbstractCodeCleanupService(ICodeFixService codeFixService)
             return document;
         }
 
+        private async Task<Document> ApplyThirdPartyCodeFixesAsync(Document document, IProgressTracker progressTracker, CodeActionOptions options, CancellationToken cancellationToken)
+        {
+            var tree = await document.GetRequiredSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            var range = new TextSpan(0, tree.Length);
+
+            // Compute diagnostics for everything that is not an IDE analyzer
+            var allDiagnostics = (await _diagnosticService.GetDiagnosticsForSpanAsync(document, range,
+                shouldIncludeDiagnostic: static diagnosticId => !(IDEDiagnosticIdToOptionMappingHelper.IsKnownIDEDiagnosticId(diagnosticId)),
+                includeSuppressedDiagnostics: false,
+                cancellationToken: cancellationToken).ConfigureAwait(false));
+
+            // ensure more than just compiler diagnostics were returned
+            var diagnostics = allDiagnostics.WhereAsArray(d => !IsCompilerDiagnostic(d.Id));
+            if (!diagnostics.Any())
+            {
+                return document;
+            }
+
+            foreach (var diagnosticId in diagnostics.SelectAsArray(static d => d.Id).Distinct())
+            {
+                // Apply codefixes for diagnostics with a severity of warning or higher
+                document = await _codeFixService.ApplyCodeFixesForSpecificDiagnosticIdAsync(document, diagnosticId, DiagnosticSeverity.Warning, progressTracker, options, cancellationToken).ConfigureAwait(false);
+            }
+
+            return document;
+
+            static bool IsCompilerDiagnostic(string errorId)
+            {
+                if (!string.IsNullOrEmpty(errorId) && errorId.Length > 2)
+                {
+                    var prefix = errorId.Substring(0, 2);
+                    if (prefix.Equals("CS", StringComparison.OrdinalIgnoreCase) || prefix.Equals("BC", StringComparison.OrdinalIgnoreCase))
+                    {
+                        var suffix = errorId.Substring(2);
+                        return int.TryParse(suffix, out _);
+                    }
+                }
```

Comment:
@mavasani can we do this another way?  For example, isn't the compiler analyzer a well known one?

Reply:
AFAICT we typically look at the tags om the diagnostic https://github.com/dotnet/roslyn/blob/fd441234f5a516fae447785f860001e99c05cf71/src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs#L396

I can update the diagnostic service to also give us the diagnostic descriptor and/or just the tags so this helper is not necessary

Reply:
yes please.  this seems not great.  can you check with manish about a better way?  i would also be fine passing a boolean into the lower layers so that they then skip the compiler analyzer.